### PR TITLE
Feature: Add alternative setting for right-click close window option to exclude pinned windows

### PR DESCRIPTION
**Pull request by ShreksHellraiser**
## Motivation / Problem

The right-click to close windows option is a great gameplay improvement, however it makes it very easy and common to close windows you don't intend to. 

## Description

My solution to this issue is to change the right-click to close windows setting to be a dropdown rather than a simple toggle, with a third option to exclude pinned windows. When set to this third mode, simply pin a window and right-click will no longer close the window.

## Limitations

Since I changed the type of the config option from a `bool` to a `uint8` the game will give you a warning upon launching if you last saved settings from a version that did not make this change. This window can be easily dismissed and as far as I can tell there are no negative effects from this.

I've been looking into solutions for this, but modifying the setting loading code to smoothly change this format to the new type is not a satisfactory solution.

I'm also not very satisfied with my if condition in `window.cpp`, it's quite long, but I'm not sure of a better way to split it.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

**Comment by LC-Zorg** [comment]
Player #623071 review: For my way of navigating the game interface, this would be a redundant option, although I do occasionally close a window I wanted to keep open. As the new behavior would be an added option, not a replacement, I have nothing against. Ok, just one thing: the red/green button is more visible, but I'm not convinced that's very important. However, I would have two suggestions.

1. Please make right-click closing the default setting. I fully agree that this is a great gameplay improvement. It makes the game so much easier, I can't see any compelling arguments against it, so why shouldn't new players be able to take advantage of it? :)

2. In the case of setting where clicking on a pinned window doesn't close it, maybe a good solution would be to allow it to be closed using Ctrl + click?

**Comment by MasonGulu** [comment]
I don't really have any authority to change default behavior, I'm just adding another option to help with accessibility.

I could add a modifier key to close the window, but I feel like at that point it's just as easy to unpin the window and close it. Adding a modifier key also increases the complexity of documenting this, would I hardcore the modifier, or would I use an existing keybind?

I'd like to receive some more input before trying to learn how user keyboard and keybind input works. But thank you for the feedback!

File: src/right_click_type.h
Comment (PeterN) [inline src/right_click_type.h]: This enum should either include a prefix `RCC_` on each value or be declared as `enum class` and use Pascal case values.

File: src/settings_type.h
Comment (PeterN) [inline src/settings_type.h]: Use the declared enum type?

File: src/table/settings/gui_settings.ini
Comment (PeterN) [inline src/table/settings/gui_settings.ini]: You might be able to use the enum values here instead of integers, not sure though.

File: src/window.cpp
Comment (PeterN) [inline src/window.cpp]: Indentation

File: src/window.cpp
Comment (PeterN) [inline src/window.cpp]: Spurious double-space before first &&

**Review by PeterN — DISMISSED** [review]

File: src/settings_type.h
Comment (glx22) [inline src/settings_type.h]:
```suggestion
	uint8  right_mouse_wnd_close;            ///< close window with right click
```

Comment (glx22) [reply src/settings_type.h]: And @PeterN has a point too.

File: src/settings_type.h
Comment (PeterN) [inline src/settings_type.h]: I think it still needs `: uint8` hereto force the enum type as the settings type is SLE_UINT8.

**Comment by MasonGulu** [comment]
I've changed the enum type. 

Is the setting type warning something that needs fixed/mitigated?

**Comment by 2TallTyler** [comment]
> Is the setting type warning something that needs fixed/mitigated?

I think so, yes. Most players don't know that the warning is harmless, and being greeted by a warning immediately upon opening a newly-updated game release is bound to result in bug reports, complaints, and a bad taste in the mouth of some players.

How to actually fix that, I don't know. 🤷

**Comment by MasonGulu** [comment]
I am looking into changing the setting type to `SDT_OMANY` so I can define a `load` function to handle conversion.

How do I fix the last commit name? The coding style guide doesn't mention a commit name for catching up to master.

**Comment by 2TallTyler** [comment]
Merge commits shouldn't be included. You'll need to rebase and force push.

**Comment by 2TallTyler** [comment]
@MasonGulu: Still interested in working on this? I would suggest asking in the `#openttd-development` channel in the official Discord server if you aren't sure how to update the setting to avoid the error message.

**Comment by MasonGulu** [comment]
I have resolved the warning on first launch issue by renaming the option to `right_click_wnd_close` from `right_mouse_wnd_close`. This has the added benefit of not interfering with older game versions, in cases when the config is shared between different instances.

File: src/window.cpp
Comment (2TallTyler) [inline src/window.cpp]:
```suggestion
	} else if (_settings_client.gui.right_click_wnd_close == RCC_YES_EXCEPT_STICKY && (w->flags & WF_STICKY) == 0 && (w->window_desc->flags & WDF_NO_CLOSE) == 0) {
```

File: src/window.cpp
Comment (2TallTyler) [reply src/window.cpp]: Comment matches the `else if` below, unless you mean something else?

File: src/table/settings/gui_settings.ini
Comment (2TallTyler) [reply src/table/settings/gui_settings.ini]: Was this not possible?

File: src/settings_type.h
Comment (2TallTyler) [inline src/settings_type.h]: Indentation:
```suggestion
	uint8  right_click_wnd_close;            ///< close window with right click
```

File: src/table/settings/gui_settings.ini
Comment (ShreksHellraiser) [reply src/table/settings/gui_settings.ini]: This slipped my mind when I picked this back up, this is possible and has been done in the latest commit

**Comment by 2TallTyler** [comment]
Try an interactive rebase (`git rebase -i head~2` to edit the last two commits, then `fixup` the latest commit) to squash that commit and make commit checker happy 😉.

**Review by 2TallTyler — DISMISSED** [review]
Looks good to me. Thanks for the contribution!

(Since you haven't done this before: from here it waits for another developer to look it over and either request changes, or agree that it's suitable for inclusion in OpenTTD and hit the Merge button 🙂.)

File: src/settings_type.h
Comment (TrueBrain) [inline src/settings_type.h:84]: These names are a bit unbalanced. `no-action` sounds good, but `yes` is then a bit weird. I would either expect no/yes, or some other wording for yes.

To make code more explicit, I don't mind the `RCC_NO_ACTION`. A suggestion could be `RCC_CLOSE_ALL` as the second option (and `RCC_CLOSE_EXCEPT_STICKY` for the third). But otherwise please make it `RCC_NO`. Either choice is fine by me.

The reason for this is that it might otherwise confuse future-us in a year or so, thinking that there was some special meaning here.

File: src/table/settings/gui_settings.ini
Comment (TrueBrain) [inline src/table/settings/gui_settings.ini]: This comment doesn't actually add value, funny enough. It is useful for this PR, sure, but once it is merged, it is no longer useful. Imagine if we leave everything we once removed in comments behind, the codebase would be littered with those comments :)

For this we use version-control. If someone actually wants to know these things, they can dive into the old history themselves :)

File: src/window.cpp
Comment (TrueBrain) [reply src/window.cpp]: The comment is in the wrong place. It is unclear it is a comment for the `else if` statement. Instead, if you move it inside of the body of that statement, and indent it, it becomes much more clear what is going on. So something like:

```
} else if (...) {
    /* Right-click close is enabled, but excluding pinned windows. */
   w->Close();
```

PS: missing dot at end of sentence in that comment. Please add that. Years ago we changed that sentences should end in dots; it is only older code that doesn't do that yet. You don't have to change older code; only when you touch it :)

File: src/table/settings/gui_settings.ini
Comment (TrueBrain) [inline src/table/settings/gui_settings.ini:132]: I do not like that we reset the setting of everything. As a player, that would be rather frustrating: because someone wanted to add an option, everyone has to suffer, basically ;)

But, it is not the first time we encounter this situation .. and the solution takes a tiny bit of work.

~The best example I know of, is `_old_units`. You can grep the code for that, and see how it is done. It does require a savegame bump to be handled properly, which might be a bit complicated for your first PR, so if you can't figure it out, do ask for help. But basically:~

~Keep the old setting around, and use `var` to store it in an `_old_NNN` variable. Add a new `SLV_NNN` to bump the savegame version. In `afterload.cpp`, check all savegames before the new version, and use the `_old_NNN` to write to the new `right_click_wnd_close` to correct value.~

~This way, nobody that was using "yes" will notice any difference.~

~PS: `NNN` ofcourse is some name you have to come up with. Not literally `NNN` ;)~

**Review by TrueBrain — CHANGES_REQUESTED** [review]
Sorry, I think @2TallTyler jumped the gun here. A few more things to pick up here :)

File: src/settings_type.h
Comment (TrueBrain) [reply src/settings_type.h]: This wasn't done? Any reason why it was resolved and not done? :)

File: src/table/settings/gui_settings.ini
Comment (TrueBrain) [inline src/table/settings/gui_settings.ini:143]: I thought I wrote a comment about this, but it seems either GitHub ate it, or I never pressed the little green button. Anyway, on the off chance of repeating myself:

Please change the type to `SDTC_OMANY` and add a `full` entry. See `gui.autosave` for an example. Now you can attach a list to it, that translates the integer to a name and back. That way, we don't see in the configuration:

`right_click_wnd_close = 2`, but instead `right_click_wnd_close = close-except-sticky`. This makes it much more understandable.

The list has to be in the same order as the enum entries are.

We have been doing this for all new enums; there are only a ton of old ones we haven't done yet.

File: src/table/settings/gui_settings.ini
Comment (TrueBrain) [reply src/table/settings/gui_settings.ini:132]: Okay, forget that, there is an easier way to do this. ~When you changed the type to `OMANY`, we can simply define a `load` to do conversion. Besides, my earlier story assumed we saved this setting in the savegame .. we don't. Even easier :D~

~Basically, with `OMANY`, it first tries to use the supplied list to convert the string to a number. If that fails, it will check if a `load` function is defined. Here you can convert `true` / `false` to the correct value. An example of this is `ConvertLandscape`, and a very recent PR I made to confirm this actually works, adding `ConvertAutosave`:~

https://github.com/OpenTTD/OpenTTD/pull/11143

~As you can see in that PR, it didn't work out-of-the-box, as the `load` was never connected in the `SDTC_OMANY` yet. But that is fixed wit that PR :)~

~If you follow that, the code becomes much easier. You do need to rename the variable back to `right_mouse_wnd_close` for it to work. But you do need to wait for that PR to be merged, ofc :)~

Comment (TrueBrain) [reply src/table/settings/gui_settings.ini:132]: And a follow-up: you can still rename the setting if we so like, and #11143 shows how you can convert the old value to the new value. It is a bit of code, and requires a configuration-version-bump, but with having the example from #11143 I am sure you can figure it out :D And if not, let us know!

PS: this would not require a `load` in the setting.

File: src/settings.cpp
Comment (TrueBrain) [inline src/settings.cpp]: We should reusing setting code here. Will write some code for that tomorrow. This are reminder to myself :)

File: src/lang/english.txt
Comment (TrueBrain) [inline src/lang/english.txt]: In two places it is called `pinned`, while in all other places it is called `sticky`. I think we should go for sticky, as that seems to be the term in use. What do you think?

File: src/settings.cpp
Comment (TrueBrain) [reply src/settings.cpp]: I am figure out with other developers what we exactly want here, and it seems we are narrowing down on a solution. Hang tight, I will help out with this soon :)

Comment (TrueBrain) [reply src/settings.cpp]: After a rebase, you can do something like:

```
		if (generic_version < IFV_RIGHT_CLICK_CLOSE && IsConversionNeeded(generic_ini, "gui", "right_mouse_wnd_close", "right_click_wnd_close", &old_item)) {
			auto old_value = BoolSettingDesc::ParseSingleValue(old_item->value->c_str());
			_settings_client.gui.right_click_wnd_close = old_value.value_or(false) ? RCC_YES : RCC_NO;
		}
```

Ignoring any typos I might have made :D

File: src/lang/english.txt
Comment (ShreksHellraiser) [reply src/lang/english.txt]: Code seems to call it sticky, while all strings refer to it as pinned. My code was a little inconsistent on that naming scheme so I can fix that.

Comment (TrueBrain) [reply src/lang/english.txt]: I couldn't find any reference to "pinned" in the strings. Did I miss any? (Honest question, to be clear)

Comment (ShreksHellraiser) [reply src/lang/english.txt]: Looking again I seem to have misremembered, the only reference to pinned I can find is in a comment https://github.com/OpenTTD/OpenTTD/blob/master/src/window.cpp#L3300

**Comment by MasonGulu** [comment]
That should be all my references to "pinned" swapped out for "sticky", rebased on the latest master branch, and with the config loading code provided

**Review by TrueBrain — APPROVED** [review]

**Comment by TrueBrain** [comment]
Tnx again for your work on this, and your patience while we figured out what we actually wanted :)

**Pull request merged**
